### PR TITLE
[4.0] Generate routed Modal links for iframes when not on the root

### PR DESCRIPTION
**Pull request by PhilETaylor**
Closes https://github.com/joomla/joomla-cms/issues/29998 

### Summary of Changes

The problem this resolves is the generating of links for iframes that start with a relative path eg `src="index.php.....` when the url you are on might be /index.php/this/that/

This then generates a url of `/index.php/this/that/index.php.....` when launched.

Admin modals are unaffected as they are always in the directory /administrator/ and therefore relatively a url starting href="index.php" would load from /administrator/

#29998 has some examples I struggled with today. 

This PR fixes two **frontend** modals - the Versions button when editing content and the SELECT button when selecting a new site logo in Template settings. 

### Testing Instructions

Install Joomal 4.0-dev @ 1f24603
Login to super admin
Click Install next to Blog Sample Data
Click ok in JS popup

Go to front end and

Login as super admin

click Welcome to your blog
Click cog -> click edit (to edit the article)

INSPECT the DIV before the Versions button

 another good example this fixes is #29296

### Actual result BEFORE applying this Pull Request

Note the url in the html 

```html
data-url="index.php?option=com_contenthistory&amp;view=history&amp;layout=modal&amp;tmpl=component&amp;field=jform_contenthistory&amp;item_id=com_content.article.3&amp;108a0c70130a44644e515f55efaca61f=1" 
```

```html
<div id="versionsModal" role="dialog" tabindex="-1" class="joomla-modal modal fade" data-url="index.php?option=com_contenthistory&amp;view=history&amp;layout=modal&amp;tmpl=component&amp;field=jform_contenthistory&amp;item_id=com_content.article.3&amp;108a0c70130a44644e515f55efaca61f=1" data-iframe="<iframe class=&quot;iframe&quot; src=&quot;index.php?option=com_contenthistory&amp;amp;view=history&amp;amp;layout=modal&amp;amp;tmpl=component&amp;amp;field=jform_contenthistory&amp;amp;item_id=com_content.article.3&amp;amp;108a0c70130a44644e515f55efaca61f=1&quot; name=&quot;Versions&quot; title=&quot;Versions&quot; height=&quot;100%&quot; width=&quot;100%&quot;></iframe>">
	<div class="modal-dialog modal-lg jviewport-width80">
		<div class="modal-content">
			<div class="modal-header">
			<h3 class="modal-title">Versions</h3>
				<button type="button" class="close novalidate" data-dismiss="modal" aria-label="Close">
			<span aria-hidden="true">×</span>
		</button>
	</div>
<div class="modal-body jviewport-height60">
	</div>
<div class="modal-footer">
	<button type="button" class="btn btn-secondary" data-dismiss="modal" aria-hidden="true">Close</button></div>
		</div>
	</div>
</div>
```

### Expected result AFTER applying this Pull Request

Note the url in the html 

```html
data-url="/index.php/component/contenthistory?view=history&amp;layout=modal&amp;tmpl=component&amp;field=jform_contenthistory&amp;item_id=com_content.article.3&amp;108a0c70130a44644e515f55efaca61f=1"
```

```html
<div id="versionsModal" role="dialog" tabindex="-1" class="joomla-modal modal fade" data-url="/index.php/component/contenthistory?view=history&amp;layout=modal&amp;tmpl=component&amp;field=jform_contenthistory&amp;item_id=com_content.article.3&amp;108a0c70130a44644e515f55efaca61f=1" data-iframe="<iframe class=&quot;iframe&quot; src=&quot;/index.php/component/contenthistory?view=history&amp;amp;layout=modal&amp;amp;tmpl=component&amp;amp;field=jform_contenthistory&amp;amp;item_id=com_content.article.3&amp;amp;108a0c70130a44644e515f55efaca61f=1&quot; name=&quot;Versions&quot; title=&quot;Versions&quot; height=&quot;100%&quot; width=&quot;100%&quot;></iframe>">
	<div class="modal-dialog modal-lg jviewport-width80">
		<div class="modal-content">
			<div class="modal-header">
			<h3 class="modal-title">Versions</h3>
				<button type="button" class="close novalidate" data-dismiss="modal" aria-label="Close">
			<span aria-hidden="true">×</span>
		</button>
	</div>
<div class="modal-body jviewport-height60">
	</div>
<div class="modal-footer">
	<button type="button" class="btn btn-secondary" data-dismiss="modal" aria-hidden="true">Close</button></div>
		</div>
	</div>
</div>
```

### Documentation Changes Required

none

### CODE REVIEW

**Comment by ceford** [comment]
With last nights Nightly build the Versions modal is styled with and without the patch. As I have multi-lingual sample data installed here are my links:

My before
data-url="index.php?option=com_contenthistory&amp;view=history&amp;layout=modal&amp;tmpl=component&amp;field=jform_contenthistory&amp;item_id=com_content.article.6&amp;587c449cfbb9305dd906a3bcbec547da=1" 
							
My after
data-url="/j4beta2/index.php/en/component/contenthistory?view=history&amp;layout=modal&amp;tmpl=component&amp;field=jform_contenthistory&amp;item_id=com_content.article.6&amp;587c449cfbb9305dd906a3bcbec547da=1" <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30007">issues.joomla.org/tracker/joomla-cms/30007</a>.</sub>

**Comment by PhilETaylor** [comment]
The “after” url is the correct url thus this patch is correct

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 856b214f2886d9d14a7b5f805d1c15d78da64eed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30007">issues.joomla.org/tracker/joomla-cms/30007</a>.</sub>

**Comment by PhilETaylor** [comment]
yes another good example this fixes is https://github.com/joomla/joomla-cms/issues/29296#event-3517186692

File: layouts/joomla/form/field/media.php [hunk 1]
```diff
@@ -106,6 +107,9 @@
 		. $asset . '&amp;author=' . $authorId)
 	. '&amp;fieldid={field-media-id}&amp;path=local-0:/' . $folder);
 
+// Correctly route the url to ensure its correctly using sef modes and subfolders
```
Comment (Quy) [inline layouts/joomla/form/field/media.php]:
```suggestion
// Correctly route the url to ensure it's correctly using sef modes and subfolders
```

**Comment by richard67** [comment]
Previous test is still valid because last change was only in a code comment.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 0cf8eba78e8ecc5502d5caa5184fb5ddccaae1c5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30007">issues.joomla.org/tracker/joomla-cms/30007</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30007">issues.joomla.org/tracker/joomla-cms/30007</a>.</sub>

**Comment by richard67** [comment]
Thanks!

**Pull request merged**